File: androidApp/src/main/kotlin/com/telegrammer/android/ui/auth/AuthViewModel.kt
package com.telegrammer.android.ui.auth

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.telegrammer.shared.model.AuthState
import com.telegrammer.shared.repository.AuthRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class AuthViewModel(private val authRepo: AuthRepository) : ViewModel() {

    private val _state = MutableStateFlow<AuthState>(AuthState.Unauthenticated)
    val state: StateFlow<AuthState> = _state.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    fun requestOtp(phoneNumber: String) {
        viewModelScope.launch {
            _isLoading.value = true
            _state.value = authRepo.requestOtp(phoneNumber)
            _isLoading.value = false
        }
    }

    fun verifyOtp(phoneNumber: String, code: String) {
        viewModelScope.launch {
            _isLoading.value = true
            _state.value = authRepo.verifyOtp(phoneNumber, code)
            _isLoading.value = false
        }
    }
}

File: iosApp/iosApp/ViewModels/AuthViewModel.swift
import SwiftUI
import shared

class AuthViewModel: ObservableObject {
    @Published var isLoading = false
    @Published var error: String?

    private let authRepo: AuthRepository

    init(authRepo: AuthRepository) {
        self.authRepo = authRepo
    }

    func requestOtp(phone: String) async -> Bool {
        await MainActor.run { isLoading = true; error = nil }
        do {
            let state: AuthState = try await callSuspend { completion in
                self.authRepo.requestOtp(phoneNumber: phone) { result, error in
                    completion(result, error)
                }
            }
            await MainActor.run { isLoading = false }
            if state is AuthStateError {
                let err = state as! AuthStateError
                await MainActor.run { error = err.message }
                return false
            }
            return true
        } catch {
            await MainActor.run { self.isLoading = false; self.error = error.localizedDescription }
            return false
        }
    }

    func verifyOtp(phone: String, code: String) async -> Bool {
        await MainActor.run { isLoading = true; error = nil }
        do {
            let state: AuthState = try await callSuspend { completion in
                self.authRepo.verifyOtp(phoneNumber: phone, code: code) { result, error in
                    completion(result, error)
                }
            }
            await MainActor.run { isLoading = false }
            if state is AuthStateAuthenticated {
                return true
            }
            if state is AuthStateError {
                let err = state as! AuthStateError
                await MainActor.run { error = err.message }
            }
            return false
        } catch {
            await MainActor.run { self.isLoading = false; self.error = error.localizedDescription }
            return false
        }
    }
}
